src/order/well_founded.lean

```diff
@@ -17,69 +18,45 @@ The predicate `well_founded` is defined in the core library. In this file we pro
 and provide a few new definitions: `well_founded.min`, `well_founded.sup`, and `well_founded.succ`.
 -/
 
-variables {α : Type*}
+/-! ### Generic relation
 
-namespace well_founded
+Note that throughout this section, we give lemmas the names they'd have if the order relation were
+`<`. -/
 
-theorem not_gt_of_lt {α : Sort*} {r : α → α → Prop} (h : well_founded r) :
-  ∀ ⦃a b⦄, r a b → ¬ r b a
-| a := λ b hab hba, not_gt_of_lt hba hab
-using_well_founded { rel_tac := λ _ _, `[exact ⟨_, h⟩],
-                     dec_tac := tactic.assumption }
+variables {α β : Type*}
 
-protected theorem is_asymm {α : Sort*} {r : α → α → Prop} (h : well_founded r) : is_asymm α r :=
-⟨h.not_gt_of_lt⟩
-
-instance {α : Sort*} [has_well_founded α] : is_asymm α has_well_founded.r :=
-has_well_founded.wf.is_asymm
-
-protected theorem is_irrefl {α : Sort*} {r : α → α → Prop} (h : well_founded r) : is_irrefl α r :=
-(@is_asymm.is_irrefl α r h.is_asymm)
-
-instance {α : Sort*} [has_well_founded α] : is_irrefl α has_well_founded.r :=
-is_asymm.is_irrefl
+namespace is_well_founded
+variables (r : α → α → Prop) [H : is_well_founded α r]
 
 /-- If `r` is a well-founded relation, then any nonempty set has a minimal element
 with respect to `r`. -/
-theorem has_min {α} {r : α → α → Prop} (H : well_founded r)
-  (s : set α) : s.nonempty → ∃ a ∈ s, ∀ x ∈ s, ¬ r x a
-| ⟨a, ha⟩ := (acc.rec_on (H.apply a) $ λ x _ IH, not_imp_not.1 $ λ hne hx, hne $
+theorem has_min (s : set α) : s.nonempty → ∃ a ∈ s, ∀ x ∈ s, ¬ r x a
+| ⟨a, ha⟩ := (acc.rec_on (H.wf.apply a) $ λ x _ IH, not_imp_not.1 $ λ hne hx, hne $
   ⟨x, hx, λ y hy hyx, hne $ IH y hyx hy⟩) ha
 
-/-- A minimal element of a nonempty set in a well-founded order.
-
-If you're working with a nonempty linear order, consider defining a
-`conditionally_complete_linear_order_bot` instance via
-`well_founded.conditionally_complete_linear_order_with_bot` and using `Inf` instead. -/
-noncomputable def min {r : α → α → Prop} (H : well_founded r)
-  (p : set α) (h : p.nonempty) : α :=
-classical.some (H.has_min p h)
+/-- A minimal element of a nonempty set with respect to a well-founded relation. See also
+`well_founded_lt.min` and `well_founded_gt.max`. -/
+noncomputable def min [is_well_founded α r] (s : set α) (hs : s.nonempty) : α :=
```

Comment:
I've decided to make `s` explicit in all of these theorems, since you can't really infer `s` from a proof of `s.nonempty` a lot of the time.